### ISSUE TITLE
Bug: S2 exceptions when viewing calendar

### ISSUE THREAD

**Issue by situatedbit** [loconomics/loconomics#121]
Site: live

As a professional, when I log in and go to the calendar, I am seeing two different exceptions (depending on the browser). 

To reproduce:
Sign in as a professional
Go to the calendar (in my case, I had one appointment in the morning, then two time periods blocked off in the afternoon; https://loconomics.com/#!calendar/2017-01-12T08:00:00.000Z, user: )
After the data for today loads, I see the following exceptions, depending on browser: 

On Firefox:
![firefox](https://cloud.githubusercontent.com/assets/6644016/21910381/9941c91a-d8d0-11e6-8887-0b17b4772142.jpg)

On Chrome: 
![chrome](https://cloud.githubusercontent.com/assets/6644016/21910380/96257542-d8d0-11e6-99ee-2565cb640351.jpg)

**Comment by IagoSRL** [comment]
Just documenting:
- I saw this error at some point, but trying to reproduce it I wasn't able. I continue looking at the code anyway.
- Because of the message, is a javascript error ('undefined' don't exists on C# at the server)
- and the only place when the error may arise [is here](https://github.com/loconomics/loconomics/blob/master/app/source/js/utils/availabilityCalculation.js#L358) (there is no more references to tzStart.day). Both error messages at Chrome and Firefox are for the same underlaying error, on the same line, just a different message; the more useful and exact message would be something like "Cannot read property 'day' of the undefined 'tzStart'"
- the variable **tzStart** is created (and not modified) as result of applying a timezone conversion to a moment object:
> var tzStart = localStart.clone().tz(scheduleTimeZone);

- Moment Timezone module documentation says that a call to the *tz* method can return
> undefined if a time zone has not been set.

- Because the *tz* method is both getter and setter, I found that passing a *falsy* value (undefined, null, **empty string**, false), returns ever the result of execute the getter (that is, a string with the time zone ID or undefined if no zone defined in the object)
- and, providing any value, even intentionally a fake time zone ID, will return ever a moment object. If the time zone don't exists, resulting object will return *undefined* when calling *tz* on it without parameters.
- Clearly, seems we are not providing a value for the timeZone at the line 342, that comes for a variable created few lines before as 
> var scheduleTimeZone = weeklySchedule.timeZone();

- The weekly schedule don't contain a time zone ID, maybe
  - It didn't loaded still the data in the *weeklySchedule* object.
  - The user has not one defined still (at database, value comes from field *timeZone FROM CalendarProviderAttributes* for the userID).

**Comment by IagoSRL** [comment]
Because I suspected that the problem don't happens only on users without a time zone preference set, and wanted to be able to confirm is fixed by reproducing, I continue trying and found a clear way to reproduce it:

- Start a new session (new browser tab)
- Log-in as a professional
- Go to calendar: everything OK, no error happens
- Log out (keep browser tab open)
- Log-in again as a professional (can be the same user)
- Go to calendar: the error appears (nothing changed in the data in between).

It happens that at some 'reset' of the model or viewmodel data, everything in the 'weeklySchedule' object gets a default value (empty time zone ID, empty schedule per week day), that triggers the error and after that, the actual data gets loaded and displayed properly.

I had a solution already: to use the auto detection of timezone in the function in order to prevent the empty value and the error.

But needed to investigate further in order to be sure of the fix and found the other error that makes that situation possible (the reset on this case).

**Comment by IagoSRL** [comment]
Fixed a cache error at RemoteModel that caused this bug some times, and added the auto detect time zone at the availability calculation to fix it for the other causes and even prevents this from failing if further data loading issues happens.

Discovered, though, management of the data, cache and loading is not too clear because of the use of several computed observables that triggers loading and re-calculation, and result is that at some points performance is worse since re-calculation happens twice in short periods of time. This may needs refactor to enhance the performance, a goal for another issue.

I tested well this changes, in Firefox and Chrome, (few lines affected in the end), and pushed to master. I will close this now, but still a double check and confirmation is welcome.
